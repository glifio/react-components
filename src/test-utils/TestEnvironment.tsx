import { ReactNode } from 'react'
import { MockedProvider, MockedResponse } from '@apollo/client/testing'
import { CoinType } from '@glif/filecoin-address'
import LotusRPCEngine from '@glif/filecoin-rpc-client'
import { Environment, Network, networks } from '../services/EnvironmentProvider'

const OptionalApolloMockProvider = ({ children, mocks, withApollo }) => {
  if (withApollo || mocks.length > 0)
    return <MockedProvider mocks={mocks}>{children}</MockedProvider>

  return <>{children}</>
}

export const TestEnvironment = ({
  children,
  withApollo,
  apolloMocks
}: TestEnvironmentProps) => (
  <Environment
    coinType={CoinType.TEST}
    networkName={Network.CALIBRATION}
    nodeStatusUrl={networks[Network.CALIBRATION].nodeStatusUrl}
    graphUrl={networks[Network.CALIBRATION].graphUrl}
    graphSecure={networks[Network.CALIBRATION].graphSecure}
    lotusApiUrl={networks[Network.CALIBRATION].lotusApiUrl}
    explorerUrl='https://explorer-calibration.glif.link'
    lotusApi={
      new LotusRPCEngine({
        apiAddress: networks[Network.CALIBRATION].lotusApiUrl
      })
    }
  >
    <OptionalApolloMockProvider withApollo={withApollo} mocks={apolloMocks}>
      {children}
    </OptionalApolloMockProvider>
  </Environment>
)

type TestEnvironmentProps = {
  children: ReactNode
  withApollo?: boolean
  apolloMocks?: MockedResponse[]
}

TestEnvironment.defaultProps = {
  withApollo: false,
  apolloMocks: []
}
