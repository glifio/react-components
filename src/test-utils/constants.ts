export const WALLET_ADDRESS = 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi'
export const WALLET_ID = 't0101'

export const WALLET_ADDRESS_2 = 't1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a'
export const WALLET_ID_2 = 't0102'

export const WALLET_ADDRESS_3 = 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'
export const WALLET_ID_3 = 't0103'
