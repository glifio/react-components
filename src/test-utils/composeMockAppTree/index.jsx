/* eslint-disable react/prop-types */
import React from 'react'
import { mockWalletProviderInstance } from '../../../__mocks__/@glif/filecoin-wallet-provider'
import WalletProviderWrapper from '../../services/WalletProvider'
import { initialState as walletProviderInitialState } from '../../services/WalletProvider/state'
import { composeWalletProviderState } from './composeState'
import { TestEnvironment } from '../TestEnvironment'

jest.mock('../../services/WalletProvider')

const Index = (statePreset = 'preOnboard', options = {}) => {
  // here you can pass a walletProviderInitialState and a preset to shape the store how you want it for testing
  const initialState = composeWalletProviderState(
    options?.walletProviderInitialState || walletProviderInitialState,
    statePreset
  )

  let walletProviderCache = { ...initialState }

  const cacheWalletProviderState = state => {
    walletProviderCache = { ...state }
    return <></>
  }

  const getWalletProviderState = () => walletProviderCache

  const Tree = ({ children }) => {
    return (
      <TestEnvironment>
        <WalletProviderWrapper
          options={options}
          statePreset={statePreset}
          getState={cacheWalletProviderState}
          initialState={initialState}
        >
          {children}
        </WalletProviderWrapper>
      </TestEnvironment>
    )
  }

  return {
    Tree,
    walletProvider: mockWalletProviderInstance,
    getWalletProviderState
  }
}

/**
 * This function is a wrapper that mocks everything the filecoin app needs for testing
 *
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query
 */

export default Index
