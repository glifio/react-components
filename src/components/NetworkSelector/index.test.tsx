import { act, fireEvent, render, waitFor } from '@testing-library/react'
import { TestEnvironment } from '../../test-utils/TestEnvironment'
import { NetworkSelector } from '.'
import {
  EnvironmentProvider,
  Network,
  networks
} from '../../services/EnvironmentProvider'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')
useRouter.mockImplementation(() => ({
  query: { network: Network.CALIBRATION },
  asPath: 'https://wallet.glif.io/'
}))

const windowReassignMock = jest.fn()

describe('NetworkSelector', () => {
  test('it renders the loading state first', () => {
    const { container, getByText } = render(<NetworkSelector />)
    expect(getByText(/Loading network/)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  describe('interactions', () => {
    beforeEach(() => {
      if (!global.window) global.window = Object.create(window)
      Object.defineProperty(window, 'location', {
        value: {
          assign: windowReassignMock
        }
      })

      jest
        .spyOn(require('./useNetworkName'), 'useNetworkName')
        .mockImplementation(lotusApiAddr => {
          if (lotusApiAddr === networks.mainnet.lotusApiUrl)
            return { networkName: Network.MAINNET, error: '', loading: false }

          return { networkName: Network.CALIBRATION, error: '', loading: false }
        })

      jest
        .spyOn(require('./useNetworkStatus'), 'useNetworkStatus')
        .mockImplementation(() => {
          return { networkConnected: true, error: '' }
        })
    })

    test('it renders the network after load', async () => {
      const { container, getByText } = render(
        <TestEnvironment>
          <NetworkSelector />
        </TestEnvironment>
      )

      await waitFor(() => {
        expect(getByText(Network.CALIBRATION)).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
      })
    })

    test('it renders the network options', async () => {
      const { container, getByText, getAllByText } = render(
        <TestEnvironment>
          <NetworkSelector />
        </TestEnvironment>
      )

      await waitFor(() => {
        expect(getByText(Network.CALIBRATION)).toBeInTheDocument()
      })

      act(() => {
        fireEvent.click(getByText(Network.CALIBRATION))
      })

      expect(getAllByText(Network.CALIBRATION).length).toBe(2)
      expect(getByText(Network.MAINNET)).toBeInTheDocument()
      expect(getByText(Network.HYPERSPACE)).toBeInTheDocument()
      expect(container.firstChild).toMatchSnapshot()
    })

    test('it changes the network', async () => {
      const { getByText } = render(
        <EnvironmentProvider>
          <NetworkSelector />
        </EnvironmentProvider>
      )

      await waitFor(() => {
        expect(getByText(Network.CALIBRATION)).toBeInTheDocument()
      })

      act(() => {
        fireEvent.click(getByText(Network.CALIBRATION))
      })

      act(() => {
        fireEvent.click(getByText(Network.MAINNET))
      })

      expect(windowReassignMock).toHaveBeenCalledWith('https://wallet.glif.io/')
    })
  })
})
