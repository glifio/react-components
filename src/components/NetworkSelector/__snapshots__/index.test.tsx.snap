// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSelector interactions it renders the network after load 1`] = `
.c2 {
  position: relative;
  display: inline-block;
  width: 18px;
  height: 18px;
  border-width: 1px;
  border-style: solid;
  border-radius: 50%;
  margin: 0;
  border-color: var(--green-medium);
}

.c2:after {
  position: absolute;
  display: block;
  content: '';
  top: 50%;
  left: 50%;
  width: 11px;
  height: 11px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c2:after {
  background-color: var(--green-medium);
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-s);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .network-name {
  text-transform: capitalize;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      color="green"
    />
    <span
      class="network-name"
    >
      calibrationnet
    </span>
    <span>
      ↓
    </span>
  </div>
</div>
`;

exports[`NetworkSelector interactions it renders the network options 1`] = `
.c2 {
  position: relative;
  display: inline-block;
  width: 18px;
  height: 18px;
  border-width: 1px;
  border-style: solid;
  border-radius: 50%;
  margin: 0;
  border-color: var(--green-medium);
}

.c2:after {
  position: absolute;
  display: block;
  content: '';
  top: 50%;
  left: 50%;
  width: 11px;
  height: 11px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c2:after {
  background-color: var(--green-medium);
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-s);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .network-name {
  text-transform: capitalize;
}

.c3 {
  position: absolute;
  overflow: hidden;
  top: calc(100% + var(--space-m));
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
  border-radius: 8px;
}

.c4 {
  padding: var(--space-m) var(--space-l);
  cursor: pointer;
  outline: none;
  text-align: center;
  text-transform: capitalize;
}

.c4:not(:first-child) {
  border-top: 1px solid var(--gray-light);
}

.c4:hover,
.c4:focus {
  color: var(--purple-medium);
}

.c4.selected {
  pointer-events: none;
  background-color: var(--purple-medium);
  color: var(--white);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      color="green"
    />
    <span
      class="network-name"
    >
      calibrationnet
    </span>
    <span>
      ↑
    </span>
  </div>
  <div
    class="c3"
  >
    <div
      class="c4"
      role="button"
      tabindex="1"
    >
      mainnet
    </div>
    <div
      class="c4 selected"
      role="button"
      tabindex="2"
    >
      calibrationnet
    </div>
    <div
      class="c4"
      role="button"
      tabindex="3"
    >
      wallabynet
    </div>
  </div>
</div>
`;

exports[`NetworkSelector it renders the loading state first 1`] = `
.c2 {
  position: relative;
  display: inline-block;
  width: 18px;
  height: 18px;
  border-width: 1px;
  border-style: solid;
  border-radius: 50%;
  margin: 0;
  border-color: var(--gray-medium);
}

.c2:after {
  position: absolute;
  display: block;
  content: '';
  top: 50%;
  left: 50%;
  width: 11px;
  height: 11px;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c2:after {
  background-color: var(--gray-medium);
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-s);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .network-name {
  text-transform: capitalize;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      color="gray"
    />
    <span>
      Loading network
    </span>
    <span>
      ↓
    </span>
  </div>
</div>
`;
